work/v1/types.go

```diff
@@ -101,8 +110,61 @@ type ManifestConfigOption struct {
 	// if it is not set.
 	// +optional
 	UpdateStrategy *UpdateStrategy `json:"updateStrategy,omitempty"`
+
+	// ConditionRules defines how to set manifestwork conditions for a specific manifest.
+	// +optional
+	ConditionRules []ConditionRule `json:"conditionRules,omitempty"`
+}
+
+// +kubebuilder:validation:XValidation:rule="self.type != 'CEL' || self.condition != \"\"",message="Condition is required for CEL rules"
+type ConditionRule struct {
+	// Condition is the type of condition that is set based on this rule.
+	// Any condition is supported, but certain special conditions can be used to
+	// to control higher level behaviors of the manifestwork.
+	// If the condition is Complete, the manifest will no longer be updated once completed.
+	// Required for CEL rules, WellKnownCompletions will default to the "Complete" condition
+	// +optional
+	Condition string `json:"condition"`
+
+	// Type defines how a manifest should be evaluated for a condition.
+	// It can be CEL, or WellKnownCompletions.
+	// If the type is CEL, user should specify the celExpressions field
+	// If the type is WellKnownCompletions, certain common types in k8s.io/api will be considered
+	// completed as defined by hardcoded rules.
+	// +kubebuilder:validation:Required
+	// +required
+	Type ConditionRuleType `json:"type"`
+
+	// CelExpressions defines the CEL expressions to be evaluated for the condition.
+	// Final result is the logical AND of all expressions.
+	// +optional
```

Comment:
I am wondering whether with type WellKnownCompletions, we do not need to set any other field in this struct.
If so, we might have a nested struct and put all field not needed for WellKnownCompletions to it, e.g.

```go
type ConditionRule struct {
    Condition string `json:"condition"`
    Type ConditionRuleType `json:"type"`
   // this could be a pointer, so when type is WellKnownCompletions, this is nil
    CEL *CELRule `json:"cel"`
}

type CELRule struct {
    Expressions []string `json:"expressions"`
    Message string `json:"message"`
    ....
}
```

Reply:
allowing message override is ok, then do we expect a default message for WellKnownCompletions type?

Reply:
And it might be more consistent if the type is WellKnownCondition? so for completed condition, user should set
```
condition: Completed
type: WellKownCondition
```
It would be easier for us to extend later for other condition.

Reply:
> And it might be more consistent if the type is WellKnownCondition? so for completed condition, user should set
> 
> ```
> condition: Completed
> type: WellKownCondition
> ```
> 
> It would be easier for us to extend later for other condition.

This is interesting. My thinking with `WellKnownCompletions` was that since Completed is a specialized case, users would want to be able to set that separately from other future well known conditions.

Are you suggesting that a rule like this would select just a single rule that matches the given condition
```
- condition: Completed
  type: WellKnownCondition
```

and then a rule like this would select all well knowns?
```
- type: WellKnownCondition
```

Reply:
I am thinking condition is the required field, you could have something like
```
- condition: Completed
   type: WellKnownCondition
- condition: Degraded
   type: WellKnownCondition
```
for a certain resource.

Reply:
Ok, that makes sense to me. It does make WellKnownCondition a little more difficult to use than wellKnownStatus since you need to know what conditions are well known, but I do like that it gives individual control over each one.

Reply:
hrm I see what you mean. I think alternatively, we could have 
```
conditionRule:
  type: WellKnownConditions
```
and
```
conditionRule:
  type: CEL
  cel:
   - condition: Available
      expressions:
      - ...
      messsage: "some message"
  - condition: Completed
     expressions:
      - ...
      messsage: "some message"
```
this looks ok to me also.

Reply:
I think we do need to allow a mix of `WellKnownConditions` and custom CEL.

Also in this format we would be unable to use `listMapKey: condition` to merge multiple rules since they could each express multiple conditions. Although there's already that issue with WellKnownConditions rules being overridden by custom CEL, so the handling of multiple copies of the same condition may need to be custom anyway.

Reply:
yeah, that make sense

Reply:
How do you feel about the idea I mentioned previously? Seems like it is a nice middle ground of making things easy by default but also allowing individual customization
```
- condition: Completed
  type: WellKnownCondition  # Single well known condition
```

```
- type: WellKnownCondition  # All well known conditions
```

Reply:
question is what happens when user set
```
- type: WellKnownCondition 
- type: CEL
   condition: Completed
   expressions: ....
```

Reply:
might work if add
```
// +listType=map
// +listMapKey=type
// +listMapKey=condition
```
such that there is no duplicate setting. but need take a try,

Reply:
@bhperry would you like to add // +listMapKey=type 
and we are good to merge this.

Reply:
I don't think type should be a map key. Condition is required now, so using just that as map key ensures each condition only has one rule. If we add type then OCM has to handle the case where a condition has a WellKnown and custom CEL rule.